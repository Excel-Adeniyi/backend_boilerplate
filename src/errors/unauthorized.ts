

export class Unathorized extends Error {
    constructor(
        message: string
    ) {
        super(message)
        this.message = "Unathorized user check"
    }
}