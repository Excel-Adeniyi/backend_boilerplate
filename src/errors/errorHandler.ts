import { Request, Response } from "express";
import { DatabaseError } from "./db.errors";
import { NotFoundError } from "./notfounderror";
import { Unathorized } from "./unauthorized";




export function errorHandler(err: Error, req: Request, res: Response) {
    if(err instanceof DatabaseError){
        return res.status(400).json({success: false, message: err.message})
    }
    if(err instanceof NotFoundError){
        return res.status(404).json({success: false, message: err.message})
    }
    if(err instanceof Unathorized){
        return  res.status(401).json({success: false, message: err.message })
    }
    return res.status(500).json({success: false, message: "Internal Server Error"})
}