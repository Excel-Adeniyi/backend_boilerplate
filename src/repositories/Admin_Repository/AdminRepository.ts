import { Pool, QueryResult, RowDataPacket } from "mysql2/promise";
import { AdminDetails } from "../../entities/Admin_Entity/admin_details";
import { IAdminRepository } from "../../interfaces/Admin_Interface/i_admin_repository";
import { DatabaseError } from "../../errors/db.errors";



export class AdminRepository implements IAdminRepository {
    protected pool: Pool;

    constructor(pool: Pool) {
        this.pool = pool;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    loginAdmin(username: string, password: string): Promise<AdminDetails | RowDataPacket[] | QueryResult> {
        throw new Error("Method not implemented.");
    }
    async createAdmin(payload: AdminDetails): Promise<AdminDetails | RowDataPacket[] | QueryResult> {
        const countquery1 = 'SELECT COUNT (*) as count FROM account WHERE username = ?'
        const countquery2 = 'SELECT COUNT (*) as count FROM admin_info WHERE email = ?'
        const countquery3 = 'SELECT COUNT (*) as count FROM admin_info WHERE phone = ?'

        const insertquery1 = 'INSERT INTO account (username, password) VALUES (?,?)';
        const params1 = [payload.username, payload.password];
        const selectquery2 = 'SELECT id FROM account WHERE username = ? AND password = ? '
        const insertquery3 = 'INSERT INTO admin_info (name, phone, email, account_id) VALUES(? ,? ,? ,?)'
        const connection = await this.pool.getConnection();

        try {
            connection.beginTransaction()
            const [SELECTCOUNT] = await connection.execute(countquery1, [payload.username])
            const selectcounts = SELECTCOUNT as RowDataPacket[]
            if (selectcounts[0].count > 0) {
                connection.rollback()
                throw new DatabaseError('user already exist')
            }
            await connection.execute(insertquery1, params1);

            const [selectResults] = await connection.execute(selectquery2, params1)
            const selectedResults = selectResults as RowDataPacket[]

            const [SELECTCOUNT2] = await connection.execute(countquery2, [payload.email])
            const selectcounts2 = SELECTCOUNT2 as RowDataPacket[]
            if (selectcounts2[0].count > 0) {
                connection.rollback()
                throw new DatabaseError('email already exist')
            }
            const [SELECTCOUNT3] = await connection.execute(countquery3, [payload.phone])
            const selectcounts3 = SELECTCOUNT3 as RowDataPacket[]
            if (selectcounts3[0].count > 0) {
                connection.rollback()
                throw new DatabaseError('phone already exist')
            }
            const [insertResult2] = await connection.execute(insertquery3, [payload.name, payload.phone, payload.email, selectedResults[0].id])
            await connection.commit()
            return insertResult2 as RowDataPacket[]
        } catch (error) {
            await connection.rollback()
            throw new DatabaseError(error as string)
        } finally {
            await connection.release()
        }
    }

    async updateAdmin(id: number, password: string): Promise<AdminDetails | QueryResult> {
        const query = 'UPDATE account SET password = ? WHERE id = ?'
        const params = [password, id]
        const connection = await this.pool.getConnection();
        try {
            const [updateResults] = await connection.execute(query, params)
            await connection.commit()
            return updateResults
        } catch (error) {
            await connection.rollback()
            throw new DatabaseError(error as string)
        } finally {
            await connection.release()
        }
    }

}
