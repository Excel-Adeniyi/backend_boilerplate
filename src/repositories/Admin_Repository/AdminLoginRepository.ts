import { Pool, QueryResult, RowDataPacket } from "mysql2/promise";
import { AdminDetails } from "../../entities/Admin_Entity/admin_details";
import { AdminRepository } from "./AdminRepository";
import { DatabaseError } from "../../errors/db.errors";


export class AdminLoginRepository extends AdminRepository{
    // private pool
    constructor(pool: Pool) {
        super(pool);
        // this.pool = pool;
    }
    async loginAdmin(username: string): Promise<AdminDetails | RowDataPacket[]  | QueryResult> {
        const query = 'SELECT password FROM account WHERE username = ?'
        const params = [username]

        const connection = await this.pool.getConnection();
        try {
            const [loginResults] = await connection.execute(query, params)
            await connection.commit()
            const rows = loginResults as RowDataPacket[]
            const data = rows[0]
            return data as RowDataPacket[]
        } catch (error) {
            await connection.rollback()
            throw new DatabaseError(error as string)
        } finally {
            await connection.release()
        }
    }

    
}