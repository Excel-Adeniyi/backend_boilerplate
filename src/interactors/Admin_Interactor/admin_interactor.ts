// import { AdminDetails } from "../../entities/Admin_Entity/admin_details";
import { RowDataPacket } from "mysql2";
import { IAdminInteractor } from "../../interfaces/Admin_Interface/i_admin_interactor";
import { AdminLoginRepository } from "../../repositories/Admin_Repository/AdminLoginRepository";
import { AdminRepository } from "../../repositories/Admin_Repository/AdminRepository";




export class AdminInteractor implements IAdminInteractor {
    private repository: AdminRepository
    private loginRepository: AdminLoginRepository
    constructor(repository: AdminRepository, loginRepository: AdminLoginRepository) {
        this.repository = repository
        this.loginRepository = loginRepository
    }
    async createAdmin(username: string, password: string, name: string, phone: string, email: string) {
        const payload = {
            username, password, name, phone, email,
        }
        const data = await this.repository.createAdmin(payload)
        return data as RowDataPacket[]
    }
    async updateAdmin(id: number, password: string) {

        const data = await this.repository.updateAdmin(id, password)
        return data
    }
    async loginAdmin(username: string, password: string) {
        const data = await this.loginRepository.loginAdmin(username)
        console.log(data, password)
        return data
    }

}