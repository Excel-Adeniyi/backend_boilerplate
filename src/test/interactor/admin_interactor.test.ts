import { describe } from "@jest/globals"
import { AdminRepository } from "../../repositories/Admin_Repository/AdminRepository"
import pool from "../../config/dbConfig"
import { AdminLoginRepository } from "../../repositories/Admin_Repository/AdminLoginRepository"
import { AdminInteractor } from "../../interactors/Admin_Interactor/admin_interactor"
import { DatabaseError } from "../../errors/db.errors"



describe('Admin Interactor', () => {
    afterAll(async() => {
        await pool.end()
    })
    describe("login", () => {
        it("should be able to return data from repository", async() => {
            const adminRepo = new AdminRepository(pool)
            const loginRepo = new AdminLoginRepository(pool)
            const adminInteractor = new AdminInteractor(adminRepo, loginRepo)
            const userI = await adminInteractor.loginAdmin('testuser', 'testerpasser')
            expect(userI).not.toBeNull
        })
    })

    describe("create Admin", () =>{
        it("should be able to create admin account", async() => {
            const adminRepo = new AdminRepository(pool)
            const loginRepo = new AdminLoginRepository(pool)
            const adminInteractor = new AdminInteractor(adminRepo, loginRepo)
            try{
                const userData = await adminInteractor.createAdmin('testuserI',  'testpass', "Interactor tester", "09826352631", "1test@test.com" )
                expect(userData).toMatchObject({
                    affectedRows: 1
                })
            }catch (error) {
                if(error instanceof DatabaseError){
                    expect(error).toMatchSnapshot('[DatabaseError: DatabaseError: user already exist]')
                }
            }
        })
    })

    describe("update Admin", () => {
        it("should be able to update admin password", async() => {
            const adminRepo = new AdminRepository(pool)
            const loginRepo = new AdminLoginRepository(pool)
            const adminInteractor = new AdminInteractor(adminRepo, loginRepo)

            const recordInfo = await adminInteractor.updateAdmin(20, 'testpass2')
            expect(recordInfo).toMatchObject({
                affectedRows: 1
            })
        })
    })
})