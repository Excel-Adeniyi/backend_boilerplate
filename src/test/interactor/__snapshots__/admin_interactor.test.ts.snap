// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Interactor create Admin should be able to create admin account:  [DatabaseError: DatabaseError: user already exist] 1`] = `[DatabaseError: DatabaseError: user already exist]`;

exports[`Admin Interactor create Admin should be able to create admin account: [DatabaseError: DatabaseError: user already exist] 1`] = `[DatabaseError: DatabaseError: user already exist]`;
